avi_model_tests.ipynb

```diff
@@ -0,0 +1,13890 @@
+      ],
+      "metadata": {
+        "id": "0uCNdbIH2y8O"
+      },
+      "execution_count": 7,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "image_to_caption_model = ImageCaptioningModel(embed_dim=256, hidden_dim=512, vocab_size=len(tokenizer.vocab), num_layers=1)\n",
+        "model =image_to_caption_model.to(device)"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "CTFxa4SU23XG",
+        "outputId": "88c4e2eb-4718-4bc0-89a1-e3b2daab6e58"
+      },
+      "execution_count": null,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stderr",
+          "text": [
+            "/usr/local/lib/python3.10/dist-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.\n",
+            "  warnings.warn(\n",
+            "/usr/local/lib/python3.10/dist-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet50_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet50_Weights.DEFAULT` to get the most up-to-date weights.\n",
+            "  warnings.warn(msg)\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Researching Hyperparameters"
+      ],
+      "metadata": {
+        "id": "A2vGvLg3-23d"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "To determine what a good learning rate would be, I researched a method that would help:"
+      ],
+      "metadata": {
+        "id": "XzER47W3KQS0"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "import torch\n",
+        "import matplotlib.pyplot as plt\n",
+        "import numpy as np\n",
+        "\n",
+        "def find_learning_rate(model, train_loader, criterion, init_value=1e-7, final_value=10.0, beta=0.98):\n",
+        "    num = len(train_loader) - 1\n",
+        "    mult = (final_value / init_value) ** (1/num)\n",
+        "    lr = init_value\n",
+        "    optimizer = torch.optim.Adam(model.parameters(), lr=lr)\n",
+        "    avg_loss = 0.0\n",
+        "    best_loss = float('inf')\n",
+        "    batch_num = 0\n",
+        "    losses = []\n",
+        "    log_lrs = []\n",
+        "\n",
+        "    for data in train_loader:\n",
+        "        batch_num += 1\n",
+        "        # Get the inputs and labels\n",
+        "        inputs, labels = data['images'].to(device), data['captions'].to(device)\n",
+        "\n",
+        "        # Clear gradients\n",
+        "        optimizer.zero_grad()\n",
+        "\n",
+        "        # Forward pass\n",
```

Comment:
I would consider deleting some of these comments because they're largely self explanatory but im also good if you feel like its imperative to keep